Fix validation on Go1.16

thaJeztah opened the pull request:
relates to https://github.com/moby/moby/pull/42583
also updating the Dockerfile with the changes I had locally to test


After updating to Go 1.16, the swagger validation started to fail:

    - platform in query must be of type string: "null"
    - target in query must be of type string: "null"
    - outputs in query must be of type string: "null"
    - /containers/{id}/json.State.Error in body must be of type string: "null"
    - /containers/{id}/json.Config.MacAddress in body must be of type string: "null"
    - /containers/{id}/json.Config.Domainname in body must be of type string: "null"
    - /containers/{id}/json.Config.WorkingDir in body must be of type string: "null"

Failures above look to be for situations where example values have an empty string
set (or no "default" value is set).

This looks related to changes in the `reflect` package in Go 1.16: from the
Go 1.16 release notes: https://golang.org/doc/go1.16#reflect

> *reflect*
>
> The Zero function has been optimized to avoid allocations. Code which incorrectly
> compares the returned Value to another > Value using `==` or `DeepEqual` may get
> different results than those obtained in previous Go versions. The documentation
> for [`reflect.Value`](https://golang.org/pkg/reflect/#Value) describes how to
> compare two Values correctly.

That documentation shows:

> To compare two Values, compare the results of the `Interface` method.

```go
package main

import (
    "fmt"
    "reflect"
)

func main() {
    data := ""
    typeA := reflect.Zero(reflect.TypeOf(data))
    typeB := reflect.ValueOf(data)

    fmt.Printf("typeA: %q (value %q)\n", typeA.Kind(), typeA.String())
    fmt.Printf("typeB: %q (value %q)\n", typeB.Kind(), typeB.String())

    if reflect.DeepEqual(typeA, typeB) {
        fmt.Println("equal")
    }
}
```

Running the following code in Go1.15 and below will produce:

    typeA: "string" (value "")
    typeB: "string" (value "")

Whereas Go1.16 will produce:

    typeA: "string" (value "")
    typeB: "string" (value "")
    equal

Using the `Interface()` method gives the same discrepancy though, and for our
use, ignoring empty strings looks to be OK, so adding an exception for this.

casualjim committed:
remove apparently useless condition check

taken from https://github.com/go-openapi/validate/commit/2d89078eabe906064beb227a6df561adb6945e8a

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Update Dockerfile to build from source

This allows for easier testing of the code;

    docker build -t swagger
    docker run -it --rm -v /path/to/swagger/:/app swagger validate /app/swagger.yaml

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on vendor/github.com/go-openapi/validate/type.go:
```diff
@@ -122,7 +122,8 @@ func (t *typeValidator) Validate(data interface{}) *Result {
 	if data == nil || reflect.DeepEqual(reflect.Zero(reflect.TypeOf(data)), reflect.ValueOf(data)) {
-		if len(t.Type) > 0 && !t.Type.Contains("null") { // TODO: if a property is not required it also passes this
+		// FIXME: prevent empty strings to be considered "null" in go1.16
+		if len(t.Type) > 0 && !t.Type.Contains("null") && !t.Type.Contains("string") { // TODO: if a property is not required it also passes this
```
Fix is here (added `!t.Type.Contains("string")`

thaJeztah commented:
@kolyshkin @tianon PTAL

tianon commented:
I guess I'm a little confused why we're still on a fork -- wasn't it to fix bugs on newer Go?  Wouldn't newer releases of go-swagger likely fix those too?

kolyshkin commented:
Just for the reference, here's a history of the previous fixes:
1. https://github.com/kolyshkin/go-swagger/commit/5793aa66d4b4112c2602c716516e24710e4adbb5
2. https://github.com/kolyshkin/go-swagger/pull/1/commits/306d9e849ecc68ee5bfb12437e507a2feee8031b

kolyshkin commented:
As a side note, I just attempted to transfer this to moby, and it failed with `moby/go-swagger already exists` exists.

I still want to do a transfer @thaJeztah if you'll remove the other repo (it's probably better than to just clone it, as we'll have PR history intact.

kolyshkin commented:
It _seems_ you might also need to pick https://github.com/go-openapi/validate/pull/138/commits/2d89078eabe906064beb227a6df561adb6945e8a (see https://github.com/go-openapi/validate/issues/137)

kolyshkin commented:
> It _seems_ you might also need to pick [go-openapi/validate@2d89078](https://github.com/go-openapi/validate/commit/2d89078eabe906064beb227a6df561adb6945e8a) (see [go-openapi/validate#137](https://github.com/go-openapi/validate/issues/137))

Hmm, maybe even use it _instead_ of your fix (the first commit).

Can you give it a try @thaJeztah?

kolyshkin commented:
nit: typo in the second commit (`docekr`).

thaJeztah commented:
> I guess I'm a little confused why we're still on a fork -- wasn't it to fix bugs on newer Go? Wouldn't newer releases of go-swagger likely fix those too?

Yes, it's ugly. And we should spend some time on getting rid of it. The very TL;DR is that we use this binary/code for two purposes:

1. for validation against OpenAPI/Swagger (this fix relates to that)
2. To generate some Go types from the swagger API (and validate that the ones in our repo match the swagger file)

For `1.`, we definitely _can_ use upstream/latest (maybe we should).

For `2.`, things are more "complicated"; it (needs changes in upstream an/or how we use the code. That part is an "unthankful task", and I guess nobody has spent time (again) to look into what's needed (last attempt was by @kolyshkin).


> Hmm, maybe even use it instead of your fix (the first commit).

Oh, nice find! That's what I did originally, but I somewhat assumed it was there "for a reason", so I changed it to be more "specific".

I must admit I didn't go digging in upstream, because I assumed it would've diverged too much. Let me try to either cherry-pick that commit (or see if we can roll the dependency forward; probably not).

> nit: typo in the second commit (`docekr`).

Whoops! Nice catch. Fingers+keyboard race-condition. Will fix.

thaJeztah commented:
Updated; now took the diff from https://github.com/go-openapi/validate/commit/2d89078eabe906064beb227a6df561adb6945e8a

It didn't apply "clean" because there was a comment added before/after the line that was changed, but easy to resolve.

thaJeztah commented:
I had a look at https://github.com/moby/go-swagger, and it looks like it is a "clean" fork of upstream https://github.com/go-swagger/go-swagger. None of the branches contains commits we made that are not in upstream.

I see @kolyshkin is an admin on that repository, so I _think_ you should be able to delete it (let me know if you're not able to do that); I didn't delete the repository yet, as I thought it was best to do it at the moment this repository is moved;

<img width="943" alt="Screenshot 2021-07-02 at 12 50 49" src="https://user-images.githubusercontent.com/1804568/124264043-4d80c880-db34-11eb-8797-35a0685176ab.png">

thaJeztah commented:
For moving away, back to upstream, I recalled Kir opened https://github.com/go-swagger/go-swagger/issues/2077 with some details. I just created https://github.com/moby/moby/issues/42593 to track in moby

thaJeztah committed:
Additional fixes for Go 1.14+

Similar to 306d9e84, but the same template had more occurences.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
go generate ./generator

Commit 306d9e84 updated the template, but did not regenerate
the files, and the actual template used is embeded in generator/bindata.go

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

kolyshkin reviewed (approved):
Not sure if the first two commits are at all necessary, but otherwise this LGTM.

kolyshkin commented:
> I see @kolyshkin is an admin on that repository, so I think you should be able to delete it (let me know if you're not able to do that); I didn't delete the repository yet, as I thought it was best to do it at the moment this repository is moved;

Will take a look mid next week. Please ping me if I won't.

thaJeztah commented:
Thanks Kir! Enjoy your long weekend 🤗
